lib/auth/predicate/builtins.go

```diff
@@ -0,0 +1,361 @@
+	case float32:
+		return aT < b.(float32)
+	default:
+		return false
+	}
+}
+
+func builtinOpGT(a, b any) any {
+	return builtinOpLT(b, a)
+}
+
+func builtinOpLE(a, b any) any {
+	if reflect.TypeOf(a) != reflect.TypeOf(b) {
+		return false
+	}
+
+	switch aT := a.(type) {
+	case string:
+		return aT <= b.(string)
+	case int:
+		return aT <= b.(int)
+	case float32:
+		return aT <= b.(float32)
+	default:
+		return false
+	}
+}
+
+func builtinOpGE(a, b any) any {
+	return builtinOpLE(b, a)
+}
+
+func builtinAdd(a, b any) (any, error) {
+	if reflect.TypeOf(a) != reflect.TypeOf(b) {
+		return nil, trace.BadParameter("cannot add types: %T and %T", a, b)
+	}
+
+	switch aT := a.(type) {
+	case string:
+		return aT + b.(string), nil
+	case int:
+		return aT + b.(int), nil
+	case float32:
+		return aT + b.(float32), nil
+	default:
+		return nil, trace.BadParameter("add unsupported for type type: %T", a)
+	}
+}
+
+func builtinSub(a, b any) (any, error) {
+	if reflect.TypeOf(a) != reflect.TypeOf(b) {
+		return nil, trace.BadParameter("cannot sub types: %T and %T", a, b)
+	}
+
+	switch aT := a.(type) {
+	case int:
+		return aT - b.(int), nil
+	case float32:
+		return aT - b.(float32), nil
+	default:
+		return nil, trace.BadParameter("sub unsupported for type type: %T", a)
+	}
+}
+
+func builtinMul(a, b any) (any, error) {
+	if reflect.TypeOf(a) != reflect.TypeOf(b) {
+		return nil, trace.BadParameter("cannot mul types: %T and %T", a, b)
+	}
+
+	switch aT := a.(type) {
+	case int:
+		return aT * b.(int), nil
+	case float32:
+		return aT * b.(float32), nil
+	default:
+		return nil, trace.BadParameter("mul unsupported for type type: %T", a)
+	}
+}
+
+func builtinDiv(a, b any) (any, error) {
+	if reflect.TypeOf(a) != reflect.TypeOf(b) {
+		return nil, trace.BadParameter("cannot div types: %T and %T", a, b)
+	}
+
+	switch aT := a.(type) {
+	case int:
+		return aT / b.(int), nil
+	case float32:
+		return aT / b.(float32), nil
+	default:
+		return nil, trace.BadParameter("div unsupported for type type: %T", a)
+	}
+}
+
+func builtinXor(a, b any) (any, error) {
+	if reflect.TypeOf(a) != reflect.TypeOf(b) {
+		return nil, trace.BadParameter("cannot xor types: %T and %T", a, b)
+	}
+
+	switch aT := a.(type) {
+	case int:
+		return aT ^ b.(int), nil
+	case bool:
+		return aT != b.(bool), nil
+	default:
+		return nil, trace.BadParameter("xor unsupported for type type: %T", a)
+	}
+}
+
+func builtinSplit(a, b string) (any, error) {
+	return strings.Split(a, b), nil
+}
+
+func builtinUpper(a string) (any, error) {
+	return strings.ToUpper(a), nil
+}
+
+func builtinLower(a string) (any, error) {
+	return strings.ToLower(a), nil
+}
+
+func builtinContains(a any, b string) (any, error) {
+	switch aT := a.(type) {
+	case string:
+		return strings.Contains(aT, b), nil
+	case []string:
+		for _, s := range aT {
+			if s == b {
+				return true, nil
+			}
+		}
+
+		return false, nil
+	default:
+		return nil, trace.BadParameter("contains not valid for type: %T", a)
```

Comment:
show which types it is valid for, here and elsewhere?

Reply:
Added valid types to error messages, that said, these are designed to never be hit as they shouldn't be possible to express in the predicate DSL, and hitting them would be a bug.